aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -0,0 +1,886 @@
+  int nnz_{0};
+  std::vector<int> size_;
+
+  cusparseMatDescr_t description_{0};
+
+  csrOutput(const std::vector<int> &size) : size_{size} {
+    create_general_description_(description_);
+  }
+
+  ~csrOutput() {
+    cusparseDestroyMatDescr(description_);
+  }
+
+  int size(int index) const {
+    return size_.at(index);
+  }
+};
+
+#if IS_CUSPARSE11_AVAILABLE()
+
+// RAII guard helps to support cuSparse 11 API for `A @ B` operation
+// This generic template exists because with cuSparse the `scalar_t` type could be a double or float  
+template <class scalar_t>
+struct CusparseMatrixMultiplyOp { 
+  
+  cusparseSpGEMMDescr_t spgemmDesc;
+
+  CusparseMatrixMultiplyOp() {
+    static_assert(std::is_same<float, scalar_t>::value || std::is_same<double, scalar_t>::value,
+      "cusparse csr sparse-sparse MM only supports data type of float and double.");
+    // SpGEMM Computation
+    TORCH_CUDASPARSE_CHECK(cusparseSpGEMM_createDescr(&spgemmDesc));
+  }
+
+  ~CusparseMatrixMultiplyOp() {
+    // destroy matrix/vector descriptors
+    cusparseSpGEMM_destroyDescr(spgemmDesc);
+  }
+
+  csrOutput operator ()(
+      const csrMatrixRef<scalar_t>& A,
+      const csrMatrixRef<scalar_t>& B,
+      Tensor& output_values,
+      IntTensor& output_indices) {
+    const int A_num_rows = A.size(0);
+    const int A_num_cols = A.size(1);
+    const int A_num_nnz = A.nnz_;
+
+    const int B_num_rows = B.size(0);
+    const int B_num_cols = B.size(1);
+    const int B_num_nnz = B.nnz_;
+
+    int* dA_csrOffsets = A.csr_pointers_;
+    int* dA_columns = A.csr_indices_;
+    scalar_t* dA_values = A.csr_values_;
+
+    int* dB_csrOffsets = B.csr_pointers_;
+    int* dB_columns = B.csr_indices_;
+    scalar_t* dB_values = B.csr_values_;
+
+    cudaDataType computeType;
+    if ( std::is_same<float, scalar_t>::value ) {
+      computeType = CUDA_R_32F;
+    } else if ( std::is_same<double, scalar_t>::value) {
+      computeType = CUDA_R_64F;
+    } else {
+      TORCH_CHECK(false, "Tensor types must be either float32 or float64");
+    }
+    csrOutput out({A.size(0), B.size(1)});
+
+    out.csr_pointers_ = at::empty({out.size(0) + 1}, output_indices.options().dtype(kInt));
+
+    int* dC_csrOffsets = out.csr_pointers_.data_ptr<int>();
+    int* dC_columns = nullptr;
+    scalar_t* dC_values = nullptr; 
+
+    scalar_t alpha = 1.0f;
+    scalar_t beta = 0.0f;
+    cusparseOperation_t opA = CUSPARSE_OPERATION_NON_TRANSPOSE;
+    cusparseOperation_t opB = CUSPARSE_OPERATION_NON_TRANSPOSE;
+    
+    csrMatrixRef<scalar_t> C(
+      nullptr,
+      nullptr,
+      nullptr,
+      /*nnz*/0,
+      {A_num_rows, B_num_cols}
+    );
+
+    //--------------------------------------------------------------------------
+    // CUSPARSE APIs
+    cusparseHandle_t handle = at::cuda::getCurrentCUDASparseHandle();
+    void *dBuffer1 = NULL, *dBuffer2 = NULL;
+    size_t bufferSize1 = 0, bufferSize2 = 0;
+  
+    cusparseSpMatDescr_t matA = A.description_;
```

Comment:
Do these need to be destroyed later? Is there a guard they should be wrapped in?

Reply:
Yes, they do. These are just pointers and they come from an object `A` and each object has a guard.  There are three different classes  `csrOutput`, `csrMatrixRef` and `CusparseMatrixMultiplyOp` which implement a RAII guard in different steps of the sparse-sparse matmul operation using cuSparse.